src/Security/TokenAuthenticator.php

```diff
@@ -0,0 +1,191 @@
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\User\UserInterface;
+use Symfony\Component\Security\Core\User\UserProviderInterface;
+use Symfony\Component\Security\Guard\AbstractGuardAuthenticator;
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\Exception\HttpExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+
+/**
+ * Class TokenAuthenticator.
+ */
+class TokenAuthenticator extends AbstractGuardAuthenticator
+{
+    private $client;
+    private $cache;
+
+    private $clientId;
+    private $clientSecret;
+    private $endPoint;
+
+    /**
+     * TokenAuthenticator constructor.
+     *
+     * @param string $bindOpenplatformId
+     * @param string $bindOpenplatformSecret
+     * @param string $bindOpenplatformIntrospectionUrl
+     * @param AdapterInterface $tokenCache
+     * @param HttpClientInterface $httpClient
+     */
+    public function __construct(string $bindOpenplatformId, string $bindOpenplatformSecret, string $bindOpenplatformIntrospectionUrl, AdapterInterface $tokenCache, HttpClientInterface $httpClient)
+    {
+        $this->client = $httpClient;
+        $this->cache = $tokenCache;
+
+        $this->clientId = $bindOpenplatformId;
+        $this->clientSecret = $bindOpenplatformSecret;
+        $this->endPoint = $bindOpenplatformIntrospectionUrl;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request)
+    {
+        return $request->headers->has('authorization');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCredentials(Request $request)
+    {
+        return $request->headers->get('authorization');
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getUser($credentials, UserProviderInterface $userProvider)
+    {
+        if (null === $credentials) {
+            // The token header was empty, authentication fails with HTTP Status
+            // Code 401 "Unauthorized"
+            return null;
+        }
+
+        // Parse token information from the bearer authorization header.
+        preg_match('/Bearer\s(\w+)/', $credentials, $matches);
+        if (2 !== count($matches)) {
+            return null;
+        }
+
+        $token = $matches[1];
+
+        // Try getting item from cache.
+        $item = $this->cache->getItem($token);
+        if ($item->isHit()) {
+            return $item->get();
+        }
+
+        try {
+            $response = $this->client->request(
+                'POST',
+                $this->endPoint.'?access_token='.$token,
+                [
+                    'auth_basic' => [$this->clientId, $this->clientSecret],
+                ]
+            );
+
+            if (200 !== $response->getStatusCode()) {
+                return null;
+            }
+
+            $content = $response->getContent();
+            $data = json_decode($content);
```

Comment:
Maybe we should add check that the response is json (json_decode) may decode to FALSE if not valid json and then `false === $data->active` would faild? We should handle this bette and log information for debugging about this

Reply:
Yes I known it comes from the code I made in upload :-) sorry

Reply:
Fixed